### PR TITLE
Integration tests

### PR DESCRIPTION
**Pull request by draganm**
Ginko/Gomega integration tests.
Setup
```
$ go get github.com/onsi/ginkgo/ginkgo
$ go get github.com/onsi/gomega
```
Running:
```
$ ginkgo integration
```

One test is pending - creating topic returns wrong Content-Type ATM

### CODE REVIEW

File: integration/create_topic_test.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+			var err error
+			response, err = http.Post("http://localhost:12345/"+topicName, "text/plain", nil)
+			Expect(err).ToNot(HaveOccurred())
+			response.Body.Close()
+			response, err = http.Post("http://localhost:12345/"+topicName, "text/plain", nil)
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		AfterEach(func() {
+			response.Body.Close()
```
Comment (bictorman) [inline integration/create_topic_test.go]: usually I use this little function everywhere to errcheck deferred close https://github.com/ninibe/netlog/blob/master/util.go#L10

I should have put it in a package, but I don't want to create a package just for that one, so it's a bit repeated in a few places, you could do the same here.

**Comment by bictorman** [comment]
Looking really good!  

And looks like you also found a data race, should I handle it? or do you want to fix it yourself?

FYI I started to create a Go client in [this branch](https://github.com/ninibe/netlog/tree/client) which includes some refactoring (message package) that will conflict with this, but should be easy to fix, and this will get merged first anyway.

**Comment by draganm** [comment]
Re. race ... was actually wondering during your talk too - is the index really necessary? I mean, Segments are more or less of constant size and sorted by time. Since messages have different sizes, binary search is not feasible, but linear scan is. There is a whole rant at Kafka's design page (https://kafka.apache.org/08/design.html) about performance of linear reads. Given constant size (c) of a segment then finding the message with a given timestamp would be O(c) which is O(1).
 In theory every segment's file name could be the timestamp of the first message in it, so selecting in which file to search is relatively easy - in worst case O(log n).
Would enable us to delete the whole index managing code, meaning that the data race would be gone too :).

File: integration/create_topic_test.go (hunk 1 above)
Comment (draganm) [inline integration/create_topic_test.go]: tbh. This close failing won't damage anything hence didn't bother with checking the error ...

**Comment by bictorman** [comment]
The index has been there since the very beginning, the only reason is that I saw kafka doing it and since it seemed easy to replicate I didn’t question it. Then I started to wonder too, but I left it there assuming they knew what they where doing and over time I started to see the point.

Aside from the faster offset lookup it has some other upsides, like allowing BigLog to be completely format agnostic, and additionally it got the timestamps in it which I think is nice to keep separated from the actual data.

But where “I believe” the index is going to make the real difference is [streaming](https://github.com/ninibe/netlog/blob/master/biglog/streamer.go) (wip) 

While performance is stated as a non goal, scanning one by one is impossible to achieve any meaningful throughput the moment there’s a little bit of network latency.

Streaming works similar to how kafka actually works, the clients says: give me as much as you can up to N offsets and up to X bytes of data. And the servers sends them all at once.

The streamer just needs to look at the index to calculate what it needs to send, and pass a LimitedReader directly over the data file to the ResponseWriter. This is more efficient than loading into memory all requested data before starting to reply, maybe even to just [respond with an error](https://github.com/ninibe/netlog/blob/master/biglog/index_reader.go#L21).

Notice also that for performance messages are also batched and compressed before being written.

That being said, I’d be curious to see an alternative index-less version and compare the two, but to be a fair comparison all features need to be implemented first.

It’s be also interesting to see how a regular index file instead of a pre-allocated memory-mapped would work.

File: integration/create_topic_test.go (hunk 1 above)
Comment (bictorman) [inline integration/create_topic_test.go]: sure, here the error doesn’t matter, but the errcheck in travis is not smart enough to figure that, and the build also fails here. In this particular case you can as well do `_ = response.Body.Close()`

File: integration/create_topic_test.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
+package integration_test
+
+import (
+	"io/ioutil"
+	"net/http"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/satori/go.uuid"
```
Comment (bictorman) [inline integration/create_topic_test.go]: could we reuse the UUID lib that we have already in vendor?
`github.com/comail/go-uuid/uuid`

it's a mirror of the old UUID lib in google code

**Comment by bictorman** [comment]
FYI the data race is now fixed in master.
The scanner test also needs a tiny adjustment due to #4

**Comment by draganm** [comment]
Thanks, I've fixed the rest of the tests CI should be passing now

**Comment by draganm** [comment]
Re index. Something like tar format would IMHO remove the need for index ... so every segment file would have format
HEADER->PAYLOAD->HEADER->PAYLOAD ...
HEADER contains meta information about the payload, such as creation time and the length of the payload, making it easy to skip/seek to the next header.

**Comment by bictorman** [comment]
Great, welcome to the list of contributors :)

index: We [kind of have that format already](https://github.com/ninibe/netlog/blob/d3e1bca3eb7ad450743e74105cfdf1da6ed8726d/message.go) (except the timestamp), from the netlog side. If you want to experiment, it shouldn't be too difficult to implement an alternative streamer without the index and run some benchmarks.

If you decide to do so, I'd recommend you to base it on [this branch](https://github.com/ninibe/netlog/tree/client) where I'm slowly implementing a Go client, since there the message format has already been spin off into its own package and you won't have circular dependency problems.

I still think the index is going to perform much better (specially in traditional disks), take into account that you need to inform the client about what you are actually sending in the HTTP headers before you start writing the response. You could use HTTP trailers instead, but I fear that would make developing clients more difficult (many people hasn't even heard of HTTP trailers).

Maybe I'm wrong though. Or maybe with/without index could be a setting.

**Pull request merged**